src/file_pipe_log/pipe.rs

```diff
@@ -8,23 +8,78 @@ use std::sync::Arc;
 use crossbeam::utils::CachePadded;
 use fail::fail_point;
 use log::error;
-use num_traits::FromPrimitive;
 use parking_lot::{Mutex, MutexGuard, RwLock};
 
 use crate::config::Config;
 use crate::env::FileSystem;
 use crate::event_listener::EventListener;
 use crate::metrics::*;
-use crate::pipe_log::{FileBlockHandle, FileId, FileSeq, LogQueue, PipeLog};
+use crate::pipe_log::{
+    FileBlockHandle, FileId, FileSeq, LogFileContext, LogQueue, PipeLog, Version,
+};
 use crate::{perf_context, Error, Result};
 
-use super::format::{FileNameExt, Version};
+use super::format::FileNameExt;
 use super::log_file::{build_file_reader, build_file_writer, FileHandler, LogFileWriter};
 
 struct FileCollection<F: FileSystem> {
+    /// Sequence number of the first file.
     first_seq: FileSeq,
-    active_seq: FileSeq,
+    /// Sequence number of the first file that is in use.
+    first_seq_in_use: FileSeq,
     fds: VecDeque<FileHandler<F>>,
+    /// `0` => no capbility for recycling stale files
+    /// `_` => finite volume for recycling stale files
+    capacity: usize,
+}
+
+#[cfg(test)]
+impl<F: FileSystem> Default for FileCollection<F> {
+    fn default() -> Self {
+        Self {
+            first_seq: 0,
+            first_seq_in_use: 0,
+            fds: VecDeque::new(),
+            capacity: 0,
+        }
+    }
+}
+
+impl<F: FileSystem> FileCollection<F> {
+    /// Recycle the first obsolete(stale) file and renewed with new FileId.
+    ///
+    /// Attention please, the recycled file would be automatically `renamed` in
+    /// this func.
+    pub fn recycle_one_file(&mut self, file_system: &F, dir_path: &str, dst_fd: FileId) -> bool {
+        if self.capacity == 0 || self.first_seq >= self.first_seq_in_use {
+            return false;
+        }
+        let mut ret = false;
+        let first_file_id = FileId {
+            queue: dst_fd.queue,
+            seq: self.first_seq,
+        };
+        if self.fds.pop_front().is_some() {
+            let src_path = first_file_id.build_file_path(dir_path); // src filepath
+            let dst_path = dst_fd.build_file_path(dir_path); // dst filepath
+            match file_system.rename(&src_path, &dst_path) {
+                Ok(_) => {
+                    // Update the first_seq
+                    self.first_seq = first_file_id.seq + 1;
+                    if self.first_seq >= self.first_seq_in_use {
+                        self.first_seq_in_use = self.first_seq;
+                    }
+                    ret = true;
+                }
+                Err(e) => {
+                    ret = false;
+                    error!("error while trying to recycle one expired file: {}", e);
+                }
+            }
+        }
+        // Only if the `rename` made sense, we could return success.
+        ret
+    }
 }
```

Comment:
According to codecov, you should add some tests that (1) rename failure during recycling (2) there's no file to recycle.